### PR TITLE
Hamming-Distance #2

### PR DESCRIPTION
**Pull request by BenSt099**
#### Description of Change

I know that there is already an implementation of the hamming-distance. My implementation operates on two Integer-arrays and it's more efficient than the normal 'for-loop'.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: misc/hamming_distance2.c [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main(void) {
+    test();  // run self-test implementations
+    return 0;
+}
```
Comment (realstealthninja) [inline misc/hamming_distance2.c:83]:
```suggestion
}

```

File: misc/hamming_distance2.c [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+/**
+ * @brief Function that calculates the Hamming distance between two Integer-arrays
+ * @param param1 int* Pointer for Hamming distance
+ * @param param2 int  length of the two arrays
+ * @param param3 int* first int-array
+ * @param param4 int* second int-array
+ * @returns Pointer to Hamming distance
+ */
+void hamming_distance(int* h_dist, int length, int* s1, int* s2) {
```
Comment (realstealthninja) [inline misc/hamming_distance2.c:24]: why not use int* as the return?
```suggestion
int* hamming_distance(int length, int* s1, int* s2) {
```

File: misc/hamming_distance2.c [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ */
+static void test() {
+
+    /* tests if h_dist is calculated correctly with even number of integers */
+    int s1[10] = {1,2,6,6,8,9,3,4,0,5}; 
+    int s2[10] = {1,2,7,6,8,9,3,2,0,5};
+    int* ptr1 = &s1[0];
+    int* ptr2 = &s2[0];
+    int h_dist;
+    hamming_distance(&h_dist, 10, ptr1, ptr2);
```
Comment (realstealthninja) [inline misc/hamming_distance2.c:61]: could you add a test case wehre one of these values are negative

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: misc/hamming_distance2.c (hunk 3 above)
Comment (BenSt099) [reply misc/hamming_distance2.c:61]: yes, sure

Comment (BenSt099) [reply misc/hamming_distance2.c]: ok i pushed two commits, one with the new test

File: misc/hamming_distance2.c [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    while(left < right){
+        dist += (int) s1[left] != s2[left];
+        dist += (int) s1[right] != s2[right];
+        ++left;
+        --right;
+    }
+
+    // if number of integer is uneven, then add hamming-distance from element in the middle
+    dist += (length % 2 != 0) * ((int) s1[(length - 1) / 2] != s2[(length - 1) / 2]);
+    return distptr;
```
[the remark this answers is not in the document]
Comment (BenSt099) [reply misc/hamming_distance2.c:47]: yeah you're right, probably it would be easier to return it directly

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: misc/hamming_distance2.c [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+#include <assert.h>  /// for assert
+#include <stdio.h>   /// for IO operations
+
+/**
+ * @brief Function that calculates the Hamming distance between two Integer-arrays
+ * @param param1 int* Pointer for Hamming distance
+ * @param param2 int  length of the two arrays
+ * @param param3 int* first int-array
+ * @param param4 int* second int-array
+ * @returns Pointer to Hamming distance
```
[the remark this answers is not in the document]
Comment (BenSt099) [reply misc/hamming_distance2.c]: thanks for telling me that!

**Review by realstealthninja — APPROVED** [review]
Very nicely done thank you so much for contributing!

**Comment by BenSt099** [comment]
You're Welcome. Thank you you for your time and your improvements.
